File: app/src/main/java/com/mrper/code23/fewk/widget/PullToRefreshStaggeredGridView.java
package com.mrper.code23.fewk.widget;


import android.content.Context;
import android.util.AttributeSet;
import android.widget.ListAdapter;

import com.etsy.android.grid.StaggeredGridView;
import com.handmark.pulltorefresh.library.PullToRefreshAdapterViewBase;
import com.handmark.pulltorefresh.library.PullToRefreshBase;
import com.mrper.code23.R;

/**
 * Created by admin on 2016/3/11.
 * 瀑布流控件
 */
public class PullToRefreshStaggeredGridView extends PullToRefreshAdapterViewBase<StaggeredGridView> {

    public PullToRefreshStaggeredGridView(Context context) {
        super(context);
    }

    public PullToRefreshStaggeredGridView(Context context, AttributeSet attrs) {
        super(context, attrs);
    }

    public PullToRefreshStaggeredGridView(Context context, Mode mode) {
        super(context, mode);
    }

    public PullToRefreshStaggeredGridView(Context context, Mode mode, PullToRefreshBase.AnimationStyle animStyle) {
        super(context, mode, animStyle);
    }

    @Override
    public PullToRefreshBase.Orientation getPullToRefreshScrollDirection() {
        return Orientation.VERTICAL;
    }

    @Override
    protected StaggeredGridView createRefreshableView(Context context, AttributeSet attrs) {
        StaggeredGridView scrollView = new StaggeredGridView(context, attrs);
        scrollView.setId(R.id.PullToRefreshMultiColumnListView);
        return scrollView;
    }

    public Object getItemAtPosition(int position){
        ListAdapter adapter = getRefreshableView().getAdapter();
        if(adapter == null) return null;
        return adapter.getItem(position);
    }

}

File: app/src/main/java/com/mrper/code23/fewk/widget/PullToRefreshStaggeredGridView.kt
package com.mrper.code23.fewk.widget


import android.content.Context
import android.util.AttributeSet
import com.etsy.android.grid.StaggeredGridView
import com.handmark.pulltorefresh.library.PullToRefreshAdapterViewBase
import com.handmark.pulltorefresh.library.PullToRefreshBase
import com.mrper.code23.R
import com.mrper.code23.data.adapter.DemoAdapter

/**
 * Created by admin on 2016/3/11.
 * 瀑布流控件
 */
class PullToRefreshStaggeredGridView : PullToRefreshAdapterViewBase<StaggeredGridView> {

    constructor(context: Context) : super(context)

    constructor(context: Context, attrs: AttributeSet) : super(context, attrs)

    constructor(context: Context, mode: PullToRefreshBase.Mode) : super(context, mode)

    constructor(context: Context, mode: PullToRefreshBase.Mode, animStyle: PullToRefreshBase.AnimationStyle) : super(context, mode, animStyle)

    override fun getPullToRefreshScrollDirection(): PullToRefreshBase.Orientation = PullToRefreshBase.Orientation.VERTICAL

    override fun createRefreshableView(context: Context, attrs: AttributeSet): StaggeredGridView {
        val scrollView = StaggeredGridView(context, attrs)
        scrollView.id = R.id.PullToRefreshMultiColumnListView
        return scrollView
    }

    fun getItemAtPosition(position: Int): Any? {
        val adapter = refreshableView.adapter ?: return null
        return adapter.getItem(position)
    }

    /** 滑动大顶部  */
    fun smoothScroll2Top() {
        val contentView = refreshableView
        contentView.resetToTop()
        val adapter = contentView.adapter as? DemoAdapter?
        adapter?.notifyDataSetChanged()
    }

}
